### ISSUE TITLE
c_link_args not used when detecting linker type

### ISSUE THREAD

**Issue by mstorsjo** [mesonbuild/meson#6839]
**Describe the bug**
When detecting the linker type (`_guess_nix_linker`), it only uses `c_args` but not `c_link_args`.

**To Reproduce**
I'm using a clang cross compilation setup, with a cross file that looks like this:
```
[properties]
c_args = ['--target=armv7a-linux-gnueabihf', '--sysroot=/home/martin/gcc-linaro-7.3.1-2018.05-x86_64_arm-linux-gnueabihf/arm-linux-gnueabihf/libc']
c_link_args = ['--target=armv7a-linux-gnueabihf', '--sysroot=/home/martin/gcc-linaro-7.3.1-2018.05-x86_64_arm-linux-gnueabihf/arm-linux-gnueabihf/libc', '-gcc-toolchain', '/home/martin/gcc-linaro-7.3.1-2018.05-x86_64_arm-linux-gnueabihf']
```
This fails to properly detect the linker, as clang when invoked with `c_args` doesn't find the right intended linker and ends up invoking `/usr/bin/ld` at this stage instead.

The issue can be avoided by including the `-gcc-toolchain` argument in `c_args` as well.

**Expected behavior**
I'd expect `c_link_args` to be used when performing a linker test.

**system parameters**
* what `meson --version` 0.53.999

**Comment by dcbaker** [comment]
I'm working on further refining the way we represent compilers and linkers in meson right now, and this is one of the things I want to solve. I'm going to be on vacation for a few weeks, but this is on my list of things to try to get done for 0.55

**Comment by JamBalaya56562** [comment]
Thank you for the report. This is resolved: the external link args are now included in the linker-detection command, so the linker is probed with the same flags as the real link step.

```python
# mesonbuild/linkers/detect.py
# guess_nix_linker:
ldflags = env.coredata.get_external_link_args(for_machine, comp_class.language)
# guess_win_linker:
check_args += env.coredata.get_external_link_args(for_machine, comp_class.language)
```

Since `get_external_link_args()` includes the cross/native file's `<lang>_link_args` (e.g. your `-gcc-toolchain ...`), clang now selects the intended linker during detection instead of the default one.

**Verification:** `guess_nix_linker` / `guess_win_linker` in `mesonbuild/linkers/detect.py` add `env.coredata.get_external_link_args(...)` to the detection command (`get_external_link_args` is defined in `mesonbuild/coredata.py`).

This looks safe to close. Thanks for reporting it.

**Comment by bonzini** [comment]
For the record, --target is better passed in `c` and `c_ld` instead of `c_args` and `c_link_args`. But yes this is fixed by #10654.

**Comment by mstorsjo** [comment]
> For the record, --target is better passed in `c` and `c_ld` instead of `c_args` and `c_link_args`. But yes this is fixed by [#10654](https://github.com/mesonbuild/meson/pull/10654).

Yep, I've since switched to mostly passing this kind of arguments there instead. But thanks for the follow-up!
